theano/tensor/nnet/bn.py

```diff
@@ -0,0 +1,57 @@
+import theano
+from theano.scalar import Composite
+from theano.scalar import add, sub, true_div, mul
+
+
+class BNComposite(Composite):
+
+    def __init__(self, dtype):
+        x = theano.scalar.Scalar(dtype=dtype).make_variable()
+        mean = theano.scalar.Scalar(dtype=dtype).make_variable()
+        std = theano.scalar.Scalar(dtype=dtype).make_variable()
+        gamma = theano.scalar.Scalar(dtype=dtype).make_variable()
+        beta = theano.scalar.Scalar(dtype=dtype).make_variable()
+        o = add(mul(true_div(sub(x, mean), std), gamma), beta)
+        inputs = [x, mean, std, gamma, beta]
+        outputs = [o]
+        super(BNComposite, self).__init__(inputs, outputs)
+
+    def grad(self, inps, grads):
+        x, mean, std, gamma, beta = inps
+        top, = grads
+        dx = (top * gamma) / std
+        dmean = -(top * gamma) / std
+        dstd = -(top * gamma * (x - mean)) / (std * std)
+        dgamma = top * (x - mean) / std
+        return [dx, dmean, dstd, dgamma, top]
+
+
+def batch_normalization(inputs, gamma, beta, mean, std):
+    """
+    This function will build the symbolic graph for applying batch normalization
+    to a set of activations. As no intermediate representations are stored for the
+    back-propagation, this implementation lower the memory usage, however,
+    it is 5-10% slower than a naive theano implementation, as it redo
+    some foward computations for the backprop.
```

Comment:
This PR is a continuation of the idea in lasagne comment. It is the same, a big Composite with a grad implemented. We changed how to implement the grad as it was easier/faster to implement that way.

The speed difference I timmed on a matrix of size 128,256 on the GPU was that the big op was 7% to 10% then the full graph during training. But it was using less memory.

As the quantity of memory on the GPU is a big bottleneck, I would default to the version that use less memory as it significatively less memory.

There won't be optimization that introduce it manually. If someone add it, it should not be enabled by default, but only with the low_mem tag is used.

Personnaly, as this version use much less memory, I think people would prefer that even if there is a small speed loos. If you think differently, tell it.

@ballasn didn't remart the speed different on a small model on mnist.